### PR TITLE
Yellow paper updates

### PR DESCRIPTION
**Pull request by kushti**
1. Block format
2. Transaction and block validation procedures 
3. Some new tests according to validation specification

### CODE REVIEW

**Comment by catena2w** [comment]
close https://github.com/ergoplatform/ergo/issues/364

File: papers/yellow/main.tex [hunk 1]
```diff
@@ -587,6 +562,10 @@ \subsection{\Coin{} Format}
     \item{\em bytes of a script. } To see how the script is serialized, see (\knote{link to \sigm language expressions
     serialization}). The script is to be represented as register R1 by wrapping its bytes are constant array of constant
     bytes.
+    \item{\em number of assets } which box is carrying. Represented as a one-byte signed integer.
+    \item{\em assets}. A box can carry multiple assets. A single record in this field is represented as a tuple
```
Comment (catena2w) [inline papers/yellow/main.tex]: Up to 4 assets?

File: papers/yellow/main.tex [hunk 2]
```diff
@@ -748,19 +727,71 @@ \subsection{Signing A Transaction}
...
+
+\knote{Check box and transaction sizes.}
+
+Stateful checks are:
+
+\begin{itemize}
+    \item{} total input and output ergo amounts must be the same. Adding up should be done with overflow checks.
+    \item{} all the inputs have valid spending proofs. Total cost to verify all the spending proofs should be no more
+            than a limit per block~(adjustable via miners voting).
+    \item{} inputs should have no more than 16 assets collectively. An input can not contain more than 4 assets.
```
Comment (catena2w) [inline papers/yellow/main.tex]: Do we really check, that input does not contain more than 4 assets? This check looks redundant as soon as we have it in stateless check during output creation

File: papers/yellow/main.tex [hunk 3]
```diff
@@ -748,19 +727,71 @@ \subsection{Signing A Transaction}
...
+    \item{} an output can not contain more than 4 assets. All the assets amounts must be positive.
+    \item{} transaction outputs collectively could not contain more than 16 assets.
+\end{itemize}
+
+\knote{Check box and transaction sizes.}
+
+Stateful checks are:
+
+\begin{itemize}
+    \item{} total input and output ergo amounts must be the same. Adding up should be done with overflow checks.
```
Comment (catena2w) [inline papers/yellow/main.tex]: I think first check should looks like `All inputs refers to existing boxes in UTXO`.

File: papers/yellow/main.tex [hunk 4]
```diff
@@ -748,19 +727,71 @@ \subsection{Signing A Transaction}
...
+\end{itemize}
 
 \subsection{Unified Transactions}
 
+\subsection{Full-Block Validation Rules}
+
+Below are rules for block validation when a node is verifying transactions ($VerifyTransactions = 1$)
+
+\begin{itemize}
+    \item{} every transaction in a block is referencing to inputs from the UTXO set or created by previous transactions
```
Comment (catena2w) [inline papers/yellow/main.tex]: This description allows double spending - 2 transactions that reference to the same input from UTXO.

File: papers/yellow/main.tex [hunk 5]
```diff
@@ -748,19 +727,71 @@ \subsection{Signing A Transaction}
...
+            than a limit per block~(adjustable via miners voting).
+    \item{} inputs should have no more than 16 assets collectively. An input can not contain more than 4 assets.
+    \item{} for each kind of asset in the outputs, total output amount for the asset should be no more than the total
+            input amount, if the asset identifier is not equals to identifier of the first input; otherwise, the total
+            output amount must be positive. The latter case corresponds to issuance of a new asset, while the former
+            sets asset preservation rule. Please note that for total input amount of an asset we do not require for
+            strict equality of the input and output amounts: the output amount could be less than input amount~(or zero).
+\end{itemize}
 
 \subsection{Unified Transactions}
```
Comment (catena2w) [inline papers/yellow/main.tex]: Let's write some notes, what is this subsection about

File: papers/yellow/main.tex (hunk 1 above)
Comment (kushti) [reply papers/yellow/main.tex]: @catena2w why this detail is important in serialization description ?

File: papers/yellow/main.tex (hunk 2 above)
Comment (kushti) [reply papers/yellow/main.tex]: @catena2w yes, we actually do this check in fillAssetsMap()

File: papers/yellow/main.tex (hunk 4 above)
Comment (kushti) [reply papers/yellow/main.tex]: @catena2w double spending should be prevented during UTXO set tree update (or checking UTXO set transformation proof). But currently double-spending prevention is actually broken, see https://github.com/ergoplatform/ergo/issues/387

**Comment by kushti** [comment]
@catena2w I think the PR is ready for another pass!

**Pull request merged**